src/core/primitives.rs

```diff
@@ -451,3 +451,52 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Attempts opening the Covercrypt encapsulation using the given USK. Returns
+/// the encapsulated key and associated rights upon success, otherwise returns `None`.
+pub fn full_decaps(
+    usk: &UserSecretKey,
+    encapsulation: &XEnc,
+) -> Result<Option<Vec<(Right, Secret<SHARED_SECRET_LENGTH>)>>, Error> {
+    // A = ⊙ _i (α_i. c_i)
+    let A = usk
+        .id
+        .iter()
+        .zip(encapsulation.c.iter())
+        .map(|(marker, trap)| trap * marker)
+        .fold(EcPoint::identity(), |mut acc, elt| {
+            acc = &acc + &elt;
+            acc
+        });
+
+    for enc in &encapsulation.encapsulations {
+        let mut rights_list: Vec<(Right, Secret<SHARED_SECRET_LENGTH>)> = Vec::new();
```

Comment:
One encapsulation should have a unique associated right, therefore I don't think you need a list here. I would however have used a list before the `enc` loop to collect all rights that allow opening an encapsulation.